junit-vintage-engine/src/main/java/org/junit/vintage/engine/VintageTestEngine.java

```diff
@@ -70,17 +86,107 @@ public void execute(ExecutionRequest request) {
+			false);
+
+		try (CloseableExecutor executor = parallelExecutionEnabled ? new ParallelExecutor(configurationParameters)
+				: new SerialExecutor()) {
+			executeAllChildren(executor, runnerExecutor, engineDescriptor);
+		}
+		catch (final InterruptedException e) {
+			throw new JUnitException("Error executing tests for engine " + getId() + ": " + e.getMessage(), e);
+		}
+
 		engineExecutionListener.executionFinished(engineDescriptor, successful());
 	}
 
-	private void executeAllChildren(RunnerExecutor runnerExecutor, TestDescriptor engineDescriptor) {
+	private void executeAllChildren(final Executor executor, RunnerExecutor runnerExecutor,
+			TestDescriptor engineDescriptor) throws InterruptedException {
+		final Collection<? extends TestDescriptor> children = engineDescriptor.getChildren();
+		final CountDownLatch latch = new CountDownLatch(children.size());
 		// @formatter:off
-		engineDescriptor.getChildren()
+		children
 				.stream()
 				.map(RunnerTestDescriptor.class::cast)
-				.forEach(runnerExecutor::execute);
+				.map(descriptor -> (Runnable)() -> {
+					runnerExecutor.execute(descriptor);
+					latch.countDown();
+				})
+				.forEach(executor::execute);
 		// @formatter:on
+		latch.await();
+	}
+
+	/**
+	 * Wrapper for {@link Executor} to allow it to be used in a try-with-resources block.
+	 */
+	@API(status = INTERNAL, since = "5.8")
+	protected interface CloseableExecutor extends Executor, AutoCloseable {
+		default public void close() throws JUnitException {
+		};
+	}
+
+	/**
+	 * {@link CloseableExecutor} that executes tasks synchronously.
+	 *
+	 * @since 5.8
+	 */
+	@API(status = INTERNAL, since = "5.8")
+	protected class SerialExecutor implements CloseableExecutor {
+		public void execute(final Runnable command) {
+			command.run();
+		}
+
+	}
+
+	/**
+	 * {@link CloseableExecutor} backed by a {@link ForkJoinPool} that executes tasks asynchronously
+	 * based on the settings in {@value #PARALLEL_CONFIG}. Clients *must* implement their own logic
+	 * to wait for tasks to complete.
+	 *
+	 * @since 5.8
+	 */
+	@API(status = INTERNAL, since = "5.8")
+	protected class ParallelExecutor implements CloseableExecutor {
+
+		private final ForkJoinPool pool;
+
+		/**
+		 * @param parameters test execution configuration
+		 */
+		public ParallelExecutor(final ConfigurationParameters parameters) {
+			final ConfigurationParameters prefixedParameters = new PrefixedConfigurationParameters(parameters,
+				PARALLEL_CONFIG);
+			final String strategyName = prefixedParameters.get("strategy").orElse("dynamic");
+			final ParallelExecutionConfigurationStrategy executionStrategy = DefaultParallelExecutionConfigurationStrategy.valueOf(
+				strategyName.toUpperCase());
+			final ParallelExecutionConfiguration executionConfiguration = executionStrategy.createConfiguration(
+				prefixedParameters);
+			pool = new ForkJoinPool(executionConfiguration.getParallelism());
```

Comment:
This should pass all config options like here:
https://github.com/junit-team/junit5/blob/1f00ec77c6581a40b435f5823964ab65c24fe171/junit-platform-engine/src/main/java/org/junit/platform/engine/support/hierarchical/ForkJoinPoolHierarchicalTestExecutorService.java#L82-L96

Reply:
I made the change to pass all the configuration options. However, I just used the default worker thread factory. Should I create a custom factory as well?